### PR TITLE
Make simple flow working

### LINKED ISSUE #1000: [Lake][DuckDB] Verify lake functionality and behavior is working as expected
### Motivation

To verify that
- The lake is working reliably, as expected
- The basic flows and tables are working reliably, as expected
- The commands and lake behavior is working as expected

We need to improve basic reliability and stability of the lake.
The basic duckdb behavior needs to be working as expected.

We should verify things are working by keeping it simple, and focusing on the `bronze_pdr_predictions table`.
I am recommending that we **ignore pdr-subcriptions, pdr-slots, and possibly other tables** so we can validate that the lake is behaving as expected.

image

### Verification - Inserting data into the lake and manipulating it
When you first start interacting with the lake, there will be a large a fetch/update step that will try to build everything into the lake. As these records are processed, we begin inserting them into our DB.
1. User should edit ppss such that the st_ts and end_ts to define their lake start & end time.
1. User should run `lake update` command to start fetching data, and fill the whole lake.
1. At any point, the user should be able to pause/cancel and resume w/o any errors
1. Show user how to set end_ts to "Now" so lake continues to update forever.

Once the lake is built, it's very likely that many records will have `null` entries as they are initially inserted into the database. **We are not worried about this for the moment.**

### Test - Cutting off the lake (dropping)

Let's first consider how our lake works. 
A certain amount of data and events arrive that need to be processed. Each time we do a run, we update a certain amount of records. 

|        | Run 1 | Run 2 | Run 3 |
|--------|-------|-------|-------|
| Time   | 1:00  | 2:00  | 3:00  |

Let's say we wanted to drop everything since Run 1. We would call our cli drop command, and get rid of that data.

`pdr lake drop 10000001 my_ppss.yaml sapphire-mainnet`

Which might be the equivalent of dropping all records since Run 1 -> End.
This would include the data from `[Run 2, Run 3]`.

The user would continue updating the lake by calling `pdr lake update`... which would refetch and rebuild `[Run 2, Run 3]`, getting the system up-to-date, and then continuing on from there,

### Verifying
We could consider that by dropping/cutting off part of the lake, all tables would have the same data cut-off/rows-dropped like below. Such that the data pipeline can resume from here, and all tables can be updated/resumed from the same "height".
```
WITH 
max_pdr_predictions AS (
  SELECT MAX(timestamp) AS max_timestamp FROM pdr_predictions
),
max_pdr_payouts AS (
  SELECT MAX(timestamp) AS max_timestamp FROM pdr_payouts
),
max_pdr_subscriptions AS (
  SELECT MAX(timestamp) AS max_timestamp FROM pdr_subscriptions
),
max_bronze_predictions AS (
  SELECT MAX(timestamp) AS max_timestamp FROM bronze_pdr_predictions
),
max_bronze_slots AS (
  SELECT MAX(timestamp) AS max_timestamp FROM bronze_pdr_slots
),
SELECT 
  'pdr_predictions' AS table_name, max_pdr_predictions.max_timestamp
FROM max_pdr_predictions
UNION ALL
SELECT 
  'pdr_payouts' AS table_name, max_pdr_payouts.max_timestamp
FROM max_pdr_payouts
UNION ALL
SELECT 
  'pdr_subscriptions' AS table_name, max_pdr_subscriptions.max_timestamp
FROM max_pdr_subscriptions
UNION ALL
SELECT 
  'bronze_pdr_predictions' AS table_name, max_bronze_predictions.max_timestamp
FROM max_bronze_predictions
UNION ALL
SELECT 
  'bronze_pdr_slots' AS table_name, max_bronze_slots.max_timestamp
FROM max_bronze_slots
```

### DoD
- [x] It's hard to break the lake
- [x] The basic lake flow is working end-to-end as expected
- [x] Raw and Bronze tables are updating as expected
- [x] Data is fetching, yielding, pausing, resuming, as expected
- [x] You are able to manipulate the lake to get the results you expected
- [x] bronze-predictions table has null records in it and that's ok
- [x] other tables (bronze-slots) and queries (update bronze-predictions) are disabled for the moment
- [x] update drop command to not use raw or etl filter and remove data from all the table

### Testing Data Pipeline Behavior
We need to verify that the basic workflows for inserting data are working. You should be able to do this step-by-step and have the lake and tables working, as expected.
- [x] Changing data feeds (BTC/ETH 5m or 1h) should not impact how we fetch data from subgraph for lake. Lake fetches everything from subgraph, filtering is done at the end.
- [x] Pausing/resuming the pipeline shouldn't cause any issues. Resuming the pipeline with `lake update` should just work.
- [x] Setting up st_ts (2024-04-01) and end_ts (now) should cause all data to be updated correctly
- [x] Changing st_ts (2024-04-01) to earlier should not cause data to backfill
- [x] Changing end_ts (2024-05-01) to earlier than max(timestamp) should not cause data to drop or anything drastic. Build command will just stop doing anything (the lake is considered full/complete).
- [x] All data should be fetched correctly and we should have all basic tables up-to-date.
- [x] pdr-slots, pdr-subscriptions, and other tables should be removed from the main etl-flow for now
- [x] There will be null records inside of pdr-predictions and pdr-slots
- [x] If we drop lake tables, we should also drop etl tables for the same st_ts -> end_ts
- [x] If we use drop CLI command all tables should be updated but CSVs should remain intact
- [x] The only way to fetch retroactively, is to change ppss.yaml and drop the raw tables.

### Core Components - Raw Table
- [x] checkpoint is identifying the right places to st_ts and end_ts
- [x] you can stop/cancel/resume/pause, and things resume correctly and reliably
- [x] the tables and records are being filled/appended correctly
- [x] inserting to duckdb starts after all of GQL + CSV has ended
- [x] inserting to duckdb cannot start/end part-way
- [x] raw tables are updated correctly
- [x] there are no duplicates in the raw table

### Core Components - ETL Table
- [x] checkpoint is identifying the right places to st_ts and end_ts [#1046](https://app.zenhub.com/workspaces/s3-predictoor-642ce98aa2ed9d00173ef51f/issues/gh/oceanprotocol/pdr-backend/1046)
- [x] you can stop/cancel/resume/pause, and things resume correctly and reliably
- [x] the tables and records are being filled/appended correctly
- [x] inserting to bronze tables starts after all of GQL + CSV + Raw Tables has ended
- [x] inserting to bronze cannot start/end part-way
- [x] bronze predictions table are updated correctly
- [x] there are no gaps in bronze_predictions table
- [x] there are no duplicates in the bronze table

### PR DESCRIPTION
**Pull request by KatunaNorbert**
Fixes #1000.

- [x] remove bronze_slots_table
- [x] fix: gql update flow breaks if(there is csv_data, but no records inside duckdb raw_table)
- [x] if <???-data> is missing from csv also remove it from the <???-table>
- [x] fix bronze <???-tables> start and end timestamps so that it takes the st_ts from <bronze-???-table> if it exists, otherwise st_ts from ppss and end_ts from the <raw-???-table>

### CODE REVIEW

File: pdr_backend/cli/cli_arguments_lake.py [hunk 1]
```diff
@@ -50,18 +49,27 @@ def __init__(self, plain_args):
...
-        elif plain_args[1] == "drop":
+            return
+
+        if plain_args[0] == "drop":
+            self.add_argument("ST", type=timestr, help="Start date yyyy-mm-dd")
+            return
+
+        if (plain_args[0] == "update") and (len(plain_args) > 3):
             self.add_argument("ST", type=timestr, help="Start date yyyy-mm-dd")
+            self.add_argument("END", type=timestr_or_now, help="End date yyyy-mm-dd")
```
Comment (idiom-bytes) [inline pdr_backend/cli/cli_arguments_lake.py:74]: as discussed, do not use extra args on update

File: pdr_backend/cli/cli_module.py [hunk 2]
```diff
@@ -188,6 +190,31 @@ def do_ohlcv(args, nested_args=None):
...
+        yaml_filename=args.PPSS_FILE,
+        network=args.NETWORK,
+        nested_override_args=nested_args,
+    )
+
+    gql_data_factory = GQLDataFactory(ppss)
+    etl = ETL(ppss, gql_data_factory)
+    etl.do_etl()
+
+
```
Comment (idiom-bytes) [inline pdr_backend/cli/cli_module.py:217]: not called this any longer

File: pdr_backend/cli/cli_module_lake.py [hunk 3]
```diff
@@ -2,39 +2,38 @@
...
     parser = LakeArgParser(args)
     parsed_args = parser.parse_args(args)
-
     func_name = f"do_lake_{parsed_args.subcommand}"
     if hasattr(parsed_args, "l2_subcommand_type"):
         func_name += f"_{parsed_args.l2_subcommand_type}"
-
     func = globals().get(func_name)
-
-    ppss = PPSS(yaml_filename=parsed_args.PPSS_FILE, network=parsed_args.NETWORK)
```
Comment (idiom-bytes) [inline pdr_backend/cli/cli_module_lake.py:30]: Calina suggested this, now you're removing it.
Inefficient work.

File: pdr_backend/lake/table.py [hunk 4]
```diff
@@ -45,32 +46,14 @@ def is_etl_table(table_name: str) -> bool:
...
-                "[while dropping raw tables] skipping non-raw table %s", table_name
-            )
-            continue
-
-        logger.info(
-            "[while dropping %s tables] drop table %s starting at %s",
-            type_filter,
-            table_name,
-            st,
-        )
```
Comment (idiom-bytes) [inline pdr_backend/lake/table.py:73]: this looks like it was done while "simplifying" things, but we did this to be explicit right now... we'll change this later after this is all completed

File: pdr_backend/lake/test/test_etl.py [hunk 5]
```diff
@@ -132,28 +134,28 @@ def test_etl_do_bronze_step(
...
+    # bronze_pdr_slots_records = pds.query_data("SELECT * FROM {}".format(table_name))
 
-    assert len(bronze_pdr_slots_records) == 4
-    assert bronze_pdr_slots_records["truevalue"].null_count() == 0
-    assert bronze_pdr_slots_records["roundSumStakes"].null_count() == 1
-    assert bronze_pdr_slots_records["source"].null_count() == 0
+    # assert len(bronze_pdr_slots_records) == 4
+    # assert bronze_pdr_slots_records["truevalue"].null_count() == 0
+    # assert bronze_pdr_slots_records["roundSumStakes"].null_count() == 1
+    # assert bronze_pdr_slots_records["source"].null_count() == 0
```
Comment (idiom-bytes) [inline pdr_backend/lake/test/test_etl.py:158]: this doesnt look good/correct

File: pdr_backend/lake/test/test_etl.py [hunk 6]
```diff
@@ -132,28 +134,28 @@ def test_etl_do_bronze_step(
...
     assert round(bronze_pdr_predictions_df["stake"][2], 3) == round(
-        _gql_datafactory_etl_payouts_df["stake"][3], 3
+        _gql_datafactory_etl_payouts_df["stake"][2], 3
     )
 
     # Assert bronze slots table is building correctly
-    table_name = get_table_name(bronze_pdr_slots_table_name)
-    bronze_pdr_slots_records = pds.query_data("SELECT * FROM {}".format(table_name))
+    # table_name = get_table_name(bronze_pdr_slots_table_name)
+    # bronze_pdr_slots_records = pds.query_data("SELECT * FROM {}".format(table_name))
```
Comment (idiom-bytes) [inline pdr_backend/lake/test/test_etl.py:153]: this doesnt look good/correct

**Comment by idiom-bytes** [comment]
@KatunaNorbert I tried to fix your text/comments so I can better understand them, and often times, I cant... 

**please, write out every detail correctly in your message**

like... what are the name of your tables... ????

```
fix bronze <???-tables> start and end timestamps so that it takes the st_ts from <bronze-???-table> if it exists, otherwise st_ts from ppss and end_ts from the <raw-???-table>
```

**Comment by idiom-bytes** [comment]
As far as I can tell, this PR mainly contains a way to reduce "lake raw update" and "lake etl update" into a single command

As explained: "We do not want to do this at the moment, since we're trying to actively validate that both of these steps are working e2e reliably, as-expected.Having separate steps let us test each part individually.

**(1) Please prepare any individual PRs to merge w/ DuckDB so we can get the DuckDB PR integrated.**

(2) close this PR for now
(3) keep track of it inside DuckDB Epic until we complete duckdb integration and we can automate this further.

**Comment by KatunaNorbert** [comment]
> @KatunaNorbert I tried to fix your text/comments so I can better understand them... like... what are the name of your tables... ??? I can't make sense of things
> 

It's not about a particular table, the logic is the same for all the tables, we don't have separate logic for each table. There is some fix that's for bronze tables and I mentioned that.

File: pdr_backend/lake/test/test_etl.py (hunk 6 above)
Comment (KatunaNorbert) [reply pdr_backend/lake/test/test_etl.py:153]: this was done do disable bronze_slots_table, check first point from PR description

Comment (idiom-bytes) [reply pdr_backend/lake/test/test_etl.py:153]: I thought you were trying to get that running here.. but I see now what is going on... I'll see what I can do to help out here so we can prepare the duckdb pr.

**Comment by idiom-bytes** [comment]
As discussed, we moved things from here to other PRs and merged them in wherever appopriate.

We're closing this PR and continuing to test/verify through issue #1000

File: pdr_backend/lake/gql_data_factory.py [hunk 7]
```diff
@@ -145,19 +151,47 @@ def _prepare_temp_table(self, table_name, st_ut, fin_ut, schema):
...
+            drop_tables_from_st(
+                PersistentDataStore(table.base_path),
+                csv_last_timestamp,
+                [get_table_name(table_name, TableType.TEMP)],
+            )
+            return
+
+        if (db_last_timestamp is None) or (
+            db_last_timestamp['max("timestamp")'][0] is None
+        ):
```
[the remark this answers is not in the document]
Comment (KatunaNorbert) [reply pdr_backend/lake/gql_data_factory.py:179]: Let's discuss this inside the specific PR, should be easier to understand

**Pull request closed without merging**